File: clients/npm/src/watch/sync.js
// On-demand sync — one scan of every local tool store, pushing new turns to the
// backend. Ported from the Python daemon's poll_once/_process/_exchanges + the
// sync_now entrypoint. This is how the Node client captures conversations WITHOUT
// a persistent daemon: `devmemory continue` (and the continue_here MCP tool)
// trigger one scoped scan of the local disk stores at session start.

import { apiCall } from "../api.js";
import { availableAdapters, key as convKey } from "./adapters.js";
import { resolveConversationProject, shouldSave } from "./project.js";
import { WatchState } from "./state.js";

const USER_CAP = 1500;
const ASSISTANT_CAP = 3000;

/**
 * Group an ordered message list into "User asked / Assistant response" blocks.
 * Consecutive same-role messages concatenate; a new user turn after an assistant
 * reply closes the previous block. Mirrors the Python daemon's `_exchanges`.
 */
export function exchanges(messages) {
  const blocks = [];
  let pendingUser = [];
  let pendingAssistant = [];

  const flush = () => {
    if (!pendingAssistant.length && !pendingUser.length) return;
    const user = pendingUser.join("\n").trim().slice(0, USER_CAP);
    const assistant = pendingAssistant.join("\n").trim().slice(0, ASSISTANT_CAP);
    const parts = [];
    if (user) parts.push(`User asked:\n${user}`);
    if (assistant) parts.push(`Assistant response:\n${assistant}`);
    if (parts.length) blocks.push(parts.join("\n\n"));
    pendingUser = [];
    pendingAssistant = [];
  };

  for (const msg of messages) {
    if (msg.role === "assistant") {
      pendingAssistant.push(msg.text);
    } else {
      if (pendingAssistant.length) flush();
      pendingUser.push(msg.text);
    }
  }
  flush();
  return blocks;
}

async function saveBlock(apiKey, project, content, sessionId) {
  const json = { project, block_type: "note", content, priority: 3 };
  if (sessionId) json.session_id = sessionId;
  const resp = await apiCall("POST", "/context", { apiKey, json });
  if (resp && resp.ok === false) throw new Error(resp.error || "save failed");
  return (resp && resp.session_id) || sessionId || null;
}

async function processConv(conv, state, apiKey, scopeSlug) {
  const k = convKey(conv);
  const already = state.savedCount(k);
  if (conv.messages.length <= already) return 0;

  const project = resolveConversationProject(conv.paths, conv.title, conv.remoteUrl);
  if (!project) return 0;
  if (scopeSlug != null && project.slug !== scopeSlug) return 0;

  const newMessages = conv.messages.slice(already);
  const blocks = exchanges(newMessages);
  if (!blocks.length) {
    // Consumed messages produced no saveable block — still advance the watermark
    // so we don't re-scan them forever.
    state.record(k, conv.messages.length, state.sessionId(k));
    return 0;
  }

  if (!shouldSave(project.slug)) return 0;

  let sessionId = state.sessionId(k);
  let saved = 0;
  for (const block of blocks) {
    const content = `[${conv.tool}] ${conv.title}\n\n${block}`;
    try {
      sessionId = (await saveBlock(apiKey, project, content, sessionId)) || sessionId;
      saved += 1;
    } catch {
      break; // network/backend/quota — stop this conv, keep the watermark honest
    }
  }

  // Advance watermark only if every block saved; else leave the unsaved tail.
  if (saved === blocks.length) state.record(k, conv.messages.length, sessionId);
  else state.record(k, already, sessionId);
  state.save();
  return saved;
}

/**
 * Run ONE scan of every local tool store and push new turns to the backend.
 * scopeSlug (optional) restricts saving to a single project; every other
 * project's watermark is left untouched. Never throws — a failed sync must never
 * break the caller (e.g. a restore). Returns the number of blocks saved.
 */
export async function syncNow(scopeSlug = null, apiKey) {
  let adapters;
  try {
    adapters = availableAdapters();
  } catch {
    return 0;
  }
  if (!adapters.length) return 0;

  const state = new WatchState();
  let total = 0;
  for (const adapter of adapters) {
    let convs;
    try {
      convs = adapter.conversations();
    } catch {
      continue; // one adapter failing must not sink the rest
    }
    for (const conv of convs) {
      try {
        total += await processConv(conv, state, apiKey, scopeSlug);
      } catch {
        /* skip a bad conversation */
      }
    }
  }
  return total;
}

File: src/devmemory/watch/sync.py
"""On-demand sync — scan local tool stores once and push new turns.

This is the daemon's poll loop reduced to a single pass, callable synchronously
from anywhere (MCP tools, CLI). It's how DevMemory captures conversations WITHOUT
a persistent background process: every tool already calls ``continue_here`` at
session start, and that call triggers one scoped sync of the local disk stores.

The daemon (``devmemory watch``) remains an optional opt-in for users who want
near-real-time push instead of capture-on-next-attach.
"""

from __future__ import annotations

from devmemory.watch.adapters import available_adapters
from devmemory.watch.client import Client, api_key
from devmemory.watch.daemon import poll_once
from devmemory.watch.state import WatchState


def sync_now(scope_slug: str | None = None, key: str | None = None) -> int:
    """Run ONE scan of every local tool store and push new turns to the backend.

    Args:
        scope_slug: If given, only conversations resolving to this project are
                    saved; every other project's watermark is left untouched, so
                    a later call (scoped or not) still captures them. On-demand
                    callers pass the project they just attached to.
        key:        API key. Falls back to DEVMEMORY_API_KEY / ~/.devmemory/api_key.

    Returns the number of blocks saved. Never raises — a failed sync must never
    break the caller (e.g. a restore). Returns 0 when there's no key, no adapter,
    or anything goes wrong mid-scan.
    """
    key = key or api_key()
    if not key:
        return 0
    adapters = available_adapters()
    if not adapters:
        return 0

    state = WatchState()
    client = Client(key)
    try:
        return poll_once(adapters, state, client, scope_slug=scope_slug)
    except Exception:  # noqa: BLE001 — on-demand sync is strictly best-effort
        return 0
    finally:
        client.close()
